trecs/models/mf.py

```diff
@@ -0,0 +1,293 @@
+        """
+        At each training timestep, we keep track of the user/item interactions.
+        """
+        user_item_ids = tuple(zip(self.users.user_vector, interactions))
+        interaction_df = pd.DataFrame(user_item_ids, columns=["user", "item"])
+        self.all_interactions = self.all_interactions.append(interaction_df, ignore_index=True)
+
+    def update_predicted_scores(self):
+        """
+        Calculates predicted scores for every user-item pair by first running an
+        implicit MF algorithm on the interaction data. A glorified wrapper function,
+        but necessary to keep the functionality in BaseRecommender working.
+        """
+        # if no interactions have happened yet, do a simple dot product
+        if self.all_interactions.size == 0:
+            super().update_predicted_scores()
+        else:
+            self.fit_mf()
+
+    def run(
+        self,
+        timesteps=50,
+        startup=False,
+        train_between_steps=False,
+        random_items_per_iter=0,
+        vary_random_items_per_iter=False,
+        repeated_items=True,
+        no_new_items=False,
+    ):
+        """
+        Just a simple wrapper so that by default, the RS does not refit the ImplicitMF model
+        at every timestep of the simulation. See the `run` method in
+        :`~models.recommender.BaseRecommender` for more details about what this does.
+        """
+        # reset interactions tracker so that interactions captured
+        # are only for the duration of this particular run
+        self.all_interactions = pd.DataFrame(columns=["user", "item"])
+        super().run(
+            timesteps,
+            startup,
+            train_between_steps,
+            random_items_per_iter,
+            vary_random_items_per_iter,
+            repeated_items,
+            no_new_items=no_new_items,
+        )
+
+    def startup_and_train(self, timesteps=50, no_new_items=True):
+        """
+        Wrapper for :`~models.recommender.BaseRecommender` `startup_and_train` method.
+        Ensures that by default, no new items are created during the startup training period.
+        """
+        super().startup_and_train(timesteps, no_new_items=no_new_items)
+
+    def fit_mf(self):
+        """
+        Run LensKit ImplicitMF training procedure to extract user/item latent
+        representations from interaction data.
+        """
+        self.model_params["features"] = self.num_latent_factors
+        model = als.ImplicitMF(**self.model_params)
+        model.fit(self.all_interactions)
+        self.als_model = model
+        # update latent representations
+        user_index, item_index = list(set(model.user_index_)), list(set(model.item_index_))
+        self.users_hat[user_index, :] = model.user_features_
+        self.items_hat[:, item_index] = model.item_features_.T
+        # update predicted scores
+        super().update_predicted_scores()
+
+    def process_new_items(self, new_items):
```

Comment:
I don't know about this. I know we discussed it, but I think if we allow new items, and use mean imputation based on the other item factors, that process shouldn't be a default behavior. It feels like the default behavior should still be to just not have new items.

Reply:
The process isn't default behavior, new items are only generated if the user specifically passes in a `Creators` object during initialization; i.e., `mf = ImplicitMF(Creators=c)`. This is how it works for all other models; if no `Creators` object is passed in at the outset, there are no new items during the simulation; otherwise, new items are created in accordance to the creators object.

Reply:
Fair enough. Just the same, I imagine you will likely be the first person to try this. When you do, I would definitely advise having a look at what the new item features are when created using imputation and compare them to some of the existing fitted values to make sure something totally wacky and unintended isn't happening!

Reply:
Totally, thanks for the heads up! Will do 😄